### PR TITLE
DEVX-2905 Normalized Page Titles

### PR DESCRIPTION
**Pull request by benhayehudi**
Currently, on NDP all page titles come out as the following: 

> Voice/voice Api > Numbers | Vonage API Developer

This PR adds a new method to the `PageTitle` presenter to normalize the page titles so that they show as the following:

> Voice/Voice API > Numbers | Vonage API Developer

### CODE REVIEW

File: lib/nexmo_developer/app/presenters/page_title.rb [hunk 1]
```diff
@@ -6,14 +6,35 @@ def initialize(product, document_title)
...
     end
   end
 
+  def normalized_product_title(product)
+    if product.include?('/')
+      product_name = product.split('/')[0]
+      product_category = product.split('/')[1]
+      product_name_normalized = product_name.split(' ').map do |phrase|
+        phrase.split('-').map do |word|
+          (word.downcase == 'api') || (word.downcase == 'sdk') ? word.upcase : word.titleize
```
Comment (mheap) [inline lib/nexmo_developer/app/presenters/page_title.rb:23]: Could we make this config driven? These rules work for the API group but other teams will need custom logic e.g.

```yaml
---
api: upcase
sdk: upcase
vbc: upcase
alwayslower: lowercase
```

Comment (benhayehudi) [reply lib/nexmo_developer/app/presenters/page_title.rb:23]: You know that's a really good idea. I'll refactor accordingly.

Comment (benhayehudi) [reply lib/nexmo_developer/app/presenters/page_title.rb]: @mheap, refactored to run as a config driven setting. I also updated the README and added a sample config for reference.

File: README.md [hunk 2]
```diff
@@ -221,6 +221,20 @@ products:
     dropdown: true
 ```
 
+5. Normalized Words List
+
+Each application could have their own list of custom words that are meant to be all UPPERCASE or all DOWNCASE, instead of title casing. Those words are stored inside `/config/normalized_words.yml`:
```
Comment (mheap) [inline README.md:226]: `DOWNCASE` being in uppercase confused me for a second here

File: lib/nexmo_developer/app/presenters/page_title.rb [hunk 3]
```diff
@@ -1,19 +1,59 @@
...
+      end
+      product = "#{product_name_normalized.join(' ')}/#{product_category_normalized.join(' ')}"
+    else
+      product = product.titleize
+    end
+    product
+  end
+
+  def normalized_word(word)
+    normalized_words = YAML.safe_load(File.open(WORD_FILE))
```
Comment (mheap) [inline lib/nexmo_developer/app/presenters/page_title.rb:47]: This can be memoized by making it an instance variable meaning that it's loaded once per boot

```
@normalized_words ||= YAML.safe_load(File.open(WORD_FILE))
```

File: lib/nexmo_developer/app/presenters/page_title.rb [hunk 4]
```diff
@@ -1,19 +1,59 @@
...
+
   def initialize(product, document_title)
     @product        = product
     @document_title = document_title
+
+    after_initialize!
+  end
+
+  def after_initialize!
+    raise 'You must provide a config/normalized_words.yml file in your documentation path.' unless File.exist?(WORD_FILE)
```
Comment (mheap) [inline lib/nexmo_developer/app/presenters/page_title.rb:12]: Can we make this file optional? Not all business units will have normalised words

File: README.md (hunk 2 above)
Comment (benhayehudi) [reply README.md:226]: Oops, that is confusing! I'll amend.

File: lib/nexmo_developer/app/presenters/page_title.rb (hunk 4 above)
Comment (benhayehudi) [reply lib/nexmo_developer/app/presenters/page_title.rb:12]: Yes, will do.

File: lib/nexmo_developer/app/presenters/page_title.rb [hunk 5]
```diff
@@ -6,14 +6,47 @@ def initialize(product, document_title)
...
+      "#{normalized_product_title(@product)} > #{@document_title} | #{default_title}"
     elsif @document_title
       "#{@document_title} | #{default_title}"
     else
       default_title
     end
   end
 
+  def normalized_product_title(product)
+    if product.include?('/')
```
Comment (fabianrbz) [inline lib/nexmo_developer/app/presenters/page_title.rb:18]: @mheap @benhayehudi  can we just add a new entry to the existing `config/products.yml` named `title` or something along those lines (we can even fall back to `titleize` if it is not present) instead?

This looks extremely complex and quite hard to follow

Comment (benhayehudi) [reply lib/nexmo_developer/app/presenters/page_title.rb:18]: I'll defer to both of you, and happy to rearrange if need be.

Comment (benhayehudi) [reply lib/nexmo_developer/app/presenters/page_title.rb:18]: I will add though that while the code above is a little bit involved (albeit not too much) it doesn't change any of the actual page title implementation details. If we add a `title` or the like to `products.yml`, we'll need to adjust how the page title is derived, adding additional logic to check for its presence and use it instead of the current formulation. In other words, both approaches (the one in this PR) and the one suggested will involve some complexity.

Comment (mheap) [reply lib/nexmo_developer/app/presenters/page_title.rb:18]: I can see arguments for both approaches. Adding a `title` entry in `products.yml` gives us the most flexibility, but also requires the end user to configure things each time a new product is added.

If this logic is only being used for page titles, let's switch to Fabian's recommendation. If it's used more generically, let's keep the upper/lower config

Comment (benhayehudi) [reply lib/nexmo_developer/app/presenters/page_title.rb:18]: that's a very good point, @mheap. Can we settle on the likelihood that users of `station` will want to use specialized lowercase or uppercase words throughout their text? I think that's a strong possibility to develop for.

File: lib/nexmo_developer/app/presenters/page_title.rb [hunk 6]
```diff
@@ -6,14 +6,26 @@ def initialize(product, document_title)
...
   end
 
+  def product_title_from_config(product)
+    config = load_config
+    item = config['products'].select{|item| item['path'] == product}[0]
+    item['name']
+  end
+
+  def load_config
+    config_path = "#{Rails.configuration.docs_base_path}/config/products.yml"
```
Comment (fabianrbz) [inline lib/nexmo_developer/app/presenters/page_title.rb:24]: @benhayehudi can you memoize this?

Comment (benhayehudi) [reply lib/nexmo_developer/app/presenters/page_title.rb:24]: 👍

File: lib/nexmo_developer/app/presenters/page_title.rb [hunk 7]
```diff
@@ -6,14 +6,26 @@ def initialize(product, document_title)
...
   end
 
+  def product_title_from_config(product)
+    config = load_config
+    item = config['products'].select{|item| item['path'] == product}[0]
+    item['name']
+  end
+
+  def load_config
+    config_path ||= "#{Rails.configuration.docs_base_path}/config/products.yml"
```
Comment (fabianrbz) [inline lib/nexmo_developer/app/presenters/page_title.rb:24]: you need to use the other syntax

```
def config
  @config ||= begin
    path ||= "#{Rails.configuration.docs_base_path}/config/products.yml"
    raise 'You must provide a config/products.yml file in your documentation path.' unless File.exist?(path)
    YAML.safe_load(File.open(path))
  end
```

File: lib/nexmo_developer/spec/fixtures/config/normalized_words.yml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+uppercase:
```
Comment (fabianrbz) [inline lib/nexmo_developer/spec/fixtures/config/normalized_words.yml:1]: we no longer need this file right? we can remove it from NDP too
